antiope-s3/src/Antiope/S3/Types.hs

```diff
@@ -31,3 +37,33 @@ instance ToText S3Uri where
 
 instance ToLogStr S3Uri where
   toLogStr s = fromString $ T.unpack $ toText s
+
+readString :: String -> RP.ReadPrec String
+readString s = do
+  remainder <- RP.look
+  if s `isPrefixOf` remainder
+    then do
+      replicateM_ (length s) RP.get
+      return s
+    else RP.pfail
+
+readWhile :: (Char -> Bool) -> RP.ReadPrec String
+readWhile f = do
+  remainder <- RP.look
+  let taken = takeWhile f remainder
+  replicateM_ (length taken) RP.get
+  return taken
+
+readBucketName :: RP.ReadPrec BucketName
+readBucketName = do
+  bucketName <- readWhile bucketNameChar
+  when (length bucketName < 3 || length bucketName > 63) RP.pfail
```

Comment:
Where does this and the lowercase, digit, dot or dash stuff come from?

Reply:
https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-s3-bucket-naming-requirements.html

Reply:
Test

Reply:
Make it easier for some future person and chuck that link in the code somewhere?